addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java

```diff
@@ -0,0 +1,606 @@
+        YamahaBridgeHandler brHandler = (YamahaBridgeHandler) getBridge().getHandler();
+        logger.debug("Input changed to " + currentInputID);
+
+        boolean includeInputWithNavigationControl = false;
+
+        for (String channelName : YamahaReceiverBindingConstants.CHANNELS_NAVIGATION) {
+            if (isLinked(grpNav(channelName))) {
+                includeInputWithNavigationControl = true;
+                break;
+            }
+        }
+
+        if (includeInputWithNavigationControl) {
+            logger.debug("navigation control requested by channel");
+            if (InputWithNavigationControl.supportedInputs.contains(currentInputID)) {
+                inputWithNavigationControl = new InputWithNavigationControl(currentInputID,
+                        brHandler.getCommunication(), this);
+                try {
+                    inputWithNavigationControl.updateNavigationState();
+                } catch (Exception e) {
+                    logger.error(e.getMessage());
+                }
+            } else {
+                logger.debug("navigation control  not supported by " + currentInputID);
+                inputWithNavigationControl = null;
+                naviState.invalidate();
+                navigationUpdated(naviState);
+            }
+        } else {
+            inputWithNavigationControl = null;
+            naviState.invalidate();
+            navigationUpdated(naviState);
+        }
+
+        boolean includeInputWithPlaybackControl = false;
+
+        for (String channelName : YamahaReceiverBindingConstants.CHANNELS_PLAYBACK) {
+            if (isLinked(grpPlayback(channelName))) {
+                includeInputWithPlaybackControl = true;
+                break;
+            }
+        }
+
+        if (includeInputWithPlaybackControl) {
+            logger.debug("playback control requested by channel");
+            if (InputWithPlayControl.supportedInputs.contains(currentInputID)) {
+                inputWithPlayControl = new InputWithPlayControl(currentInputID, brHandler.getCommunication(), this);
+                try {
+                    inputWithPlayControl.updatePlaybackInformation();
+                    inputWithPlayControl.updatePresetInformation();
+                } catch (Exception e) {
+                    logger.error(e.getMessage());
+                }
+            } else {
+                logger.debug("playback control  not supported by " + currentInputID);
+                inputWithPlayControl = null;
+                playControlState.invalidate();
+                playInfoState.invalidate();
+                playControlUpdated(playControlState);
+                playInfoUpdated(playInfoState);
+            }
+        } else {
+            inputWithPlayControl = null;
+            playControlState.invalidate();
+            playControlUpdated(playControlState);
+            playInfoUpdated(playInfoState);
+            playInfoState.invalidate();
+        }
+
```

Comment:
Method is split. But there is no good way of re-using code without introducing an interface for classes, who look similar but fulfill different purposes.